File: ts/src/day3/day3_a.ts
import * as fs from "fs";
import readline from "readline";

async function binaryDiagnostic() {
	const rl = readline.createInterface(fs.createReadStream("./input"));

	const nums: number[][] = [];
	for await (const line of rl) {
		const row: number[] = [];
		for (const i of line) {
			row.push(Number.parseInt(i));
		}
		nums.push(row);
	}

	const counter = new Array(nums[0].length).fill(0);
	for(const row of nums) {
		for(let i=0; i<row.length; i++) {
			counter[i] += row[i];
		}
	}

	const gamma: number[] = [];
	const epsilon: number[] = [];

	for(const i of counter) {
		if(i > nums.length/2) {
			gamma.push(1);
			epsilon.push(0);
		} else {
			gamma.push(0);
			epsilon.push(1);
		}

	}

	console.log(Number.parseInt(gamma.join(""), 2) * Number.parseInt(epsilon.join(""), 2));
}

(function main() {
	binaryDiagnostic();
})();

File: golang/day3/day3_a.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func binaryDiagnostic() {
	file, _ := os.Open("./input")
	scanner := bufio.NewScanner(file)

	var nums [1000][12]int

	i := 0
	for scanner.Scan() {
		var row [12]int
		for index, val := range scanner.Text() {
			num, _ := strconv.Atoi(string(val))
			row[index] = num
		}
		nums[i] = row
		i++
	}

	var counter [len(nums[0])]int
	for _, row := range nums {
		for index, val := range row {
			counter[index] += val
		}
	}

	var gammaArray [len(nums[0])]string
	var epsilonArray [len(nums[0])]string

	for index, val := range counter {
		if val > len(nums)/2 {
			gammaArray[index] = "1"
			epsilonArray[index] = "0"
		} else {
			gammaArray[index] = "0"
			epsilonArray[index] = "1"
		}
	}

	gamma, _ := strconv.ParseInt(strings.Join(gammaArray[:], ""), 2, 64)
	epsilon, _ := strconv.ParseInt(strings.Join(epsilonArray[:], ""), 2, 64)

	fmt.Println(gamma * epsilon)
}

func main() {
	binaryDiagnostic()
}
